## 异常

#### 10.1 异常
	
10.1.1 异常的概念
	
	异常: 运行时的错误
	
10.1.2 异常的处理

	目的: 发生异常的时候 程序继续运行
	
	try catch finally 
	
	Ctrl + Shift + T: 打开Open Type查找类文件 
	
	
10.1.3 throwable类
	
	大部分在java.lang包里
	
	throwable的父亲Object
	
	throwable的直接孩子: Error 和 Exeption
	
	Error类用于定义不指望程序能从其恢复过来的灾难性故障 所以程序员不需要捕捉他们
	
	如: 线程停止  虚拟机错误等
	
	Error的直接孩子: AWT Error 和  Thread Death
	
	Exeption: 编译时异常  运行时异常
	
		编译时异常:
	
			必须处理 不然程序无法编译
	
			Exeption的子类中除去RuntimeException类的其他所有类 都是编译时异常  是由特定代码抛出的 不能随便处理
			
		运行时异常:
		
			RuntimeException自己和其所有子类
			
			不处理类也可以编译
			
			重要的运行异常:
				
				ArrayIndexOutOfBoundsException  下标越界
				
				NullPointerException       对象没创建
				
				ClassCastException         类型转换错误
				
				ArithmetricException       非法算术操作  例如使用做除数
				
				IllegalArgumentException   传递给一个方法的实际参数和该方法的形式参数类型不符
				
				SecuriryException          程序执行违反安全规定的非法操作
				
				NuberFormatException       字符串转数字的异常
				
	Exeption的直接孩子: SQLException ClassNotFoundException 和 RunTimeException
	
	一个try后面可以有多个catch
	
	大异常 必须在 小异常 前面   
	
	兄弟异常先后关系都可以
	
10.1.4 throws

	throws使用在方法声明之后
	声明的是编译时异常  对于运行时的异常可以不声明
	所以编译时异常必须声明的 不然会报错
	
	public void Teacher {
		// throws使用在类的方法声明之后
		// 告诉调用shangke方法 会抛出 NullPointerException, ClassNotFoundException
		public void shangke() throws NullPointerException, ClassNotFoundException {
			try {
				
			} catch (e) {
				
				e.getMessage()      // 异常信息
				e.printStackTrace() // 调用栈信息
				throw e; // 直接抛出
			} finally {
				# code
			}
		}
	}
	

10.1.5 throw与throws总结
	
	throw:
		
		出现方法体内
		
		表示抛出某个异常
		
	throws:
		
		属于方法声明
		
		在参数列表的后面
		
		表示这个方法可能抛出的异常
		
		如果是编译时异常 调用者必须处理
		
		可以声明多个
	

10.1.6 自定义异常

	希望当一个标准异常出现时 添加一些信息 可以通过抛出你自己的异常类来实现 自定义异常
	
	代码中可以有一些出错条件 以保证区分出某种特定的异常类
	
	虽然可以从任何的标准异常派生出自己的类  但最好的策略是从Exception派生
	
	public DiyException extends Exception {
	
	}